File: templates/utils/workflow/task_manager.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
任务管理 CLI。

借鉴 Trellis task.py，提供结构化的任务查询和状态管理命令。
整合 task_parser / state_manager / dependency_checker / context_budget 的能力。

用法:
    python3 task_manager.py status          # 当前任务状态
    python3 task_manager.py --project-id <id> --project-root <dir> status
    python3 task_manager.py list            # 列出所有任务
    python3 task_manager.py next            # 下一个待执行任务
    python3 task_manager.py complete T3     # 标记完成
    python3 task_manager.py fail T3 "原因"  # 标记失败
    python3 task_manager.py deps T5         # 检查依赖
    python3 task_manager.py parallel        # 查找可并行任务
    python3 task_manager.py progress        # 进度统计
    python3 task_manager.py context-budget  # 上下文预算
"""

from __future__ import annotations

import argparse
import json
import os
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional

# Import sibling modules
sys.path.insert(0, os.path.dirname(__file__))

from context_budget import (
    detect_complexity,
    evaluate_budget_thresholds,
    generate_context_bar,
)
from dependency_checker import check_task_deps, find_parallel_groups
from path_utils import detect_project_id_from_root, get_workflow_state_path, resolve_under, validate_project_id
from state_manager import (
    calculate_progress,
    generate_progress_bar,
    read_state,
    write_state,
)
from status_utils import add_unique, get_status_emoji
from task_parser import (
    count_tasks,
    extract_constraints,
    find_next_task,
    parse_tasks_v2,
    task_to_dict,
    update_task_status_in_markdown,
)


# =============================================================================
# State & Tasks Resolution
# =============================================================================


def detect_project_id(project_root: Optional[str] = None) -> Optional[str]:
    return detect_project_id_from_root(project_root)


def detect_project_root(project_root: Optional[str] = None) -> Path:
    if project_root:
        return Path(project_root)

    config_path = Path.cwd() / ".claude" / "config" / "project-config.json"
    if config_path.is_file():
        return config_path.parents[2]
    return Path.cwd()


def resolve_plan_artifact_path(
    project_root: Path,
    artifact_ref: str,
) -> Optional[Path]:
    if not artifact_ref:
        return None

    candidate = Path(artifact_ref)
    if candidate.is_absolute():
        return candidate

    resolved = resolve_under(str(project_root), artifact_ref)
    if resolved:
        return Path(resolved)

    fallback = (project_root / artifact_ref).resolve(strict=False)
    project_root_resolved = project_root.resolve(strict=False)
    if fallback == project_root_resolved or str(fallback).startswith(str(project_root_resolved) + os.sep):
        return fallback
    return None


def resolve_state_and_tasks(
    project_id: Optional[str] = None,
    project_root: Optional[str] = None,
) -> tuple[Dict[str, Any] | None, str | None, str | None, str | None]:
    """Resolve state, state_path, task/plan content, artifact path."""
    pid = project_id or detect_project_id(project_root)
    if not pid or not validate_project_id(pid):
        return None, None, None, None

    state_path_raw = get_workflow_state_path(pid)
    if not state_path_raw:
        return None, None, None, None
    state_path = Path(state_path_raw)

    if not state_path.is_file():
        return None, None, None, None

    state = read_state(str(state_path), pid)
    resolved_project_root = detect_project_root(
        project_root or state.get("project_root")
    )
    plan_file = state.get("plan_file", "")
    artifact_path = resolve_plan_artifact_path(resolved_project_root, plan_file)

    if not artifact_path:
        return state, str(state_path), None, None

    if not artifact_path.is_file():
        return state, str(state_path), None, None

    tasks_content = artifact_path.read_text(encoding="utf-8")
    return state, str(state_path), tasks_content, str(artifact_path)


# =============================================================================
# Commands
# =============================================================================


def build_runtime_summary(state: Dict[str, Any]) -> Dict[str, Any]:
    review_status = state.get("review_status") or {}
    quality_gates = state.get("quality_gates") or {}
    return {
        "delta_tracking": state.get("delta_tracking", {}),
        "planning_gates": {
            "discussion": state.get("discussion", {}),
            "ux_design": state.get("ux_design", {}),
            "user_spec_review": review_status.get("user_spec_review", {}),
        },
        "quality_gate_summary": {
            "count": len(quality_gates.keys()),
            "passed": sorted(
                task_id for task_id, gate in quality_gates.items() if gate.get("overall_passed")
            ),
            "task_ids": sorted(quality_gates.keys()),
        },
        "unblocked": state.get("unblocked", []),
    }


def cmd_status(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """Show current workflow and task status."""
    state, state_path, tasks_content, tasks_path = resolve_state_and_tasks(
        project_id, project_root
    )
    if not state:
        return {"error": "没有活跃的工作流"}

    progress = state.get("progress", {})
    current = state.get("current_tasks", [])
    total = count_tasks(tasks_content) if tasks_content else 0
    pct = calculate_progress(
        total,
        progress.get("completed", []),
        progress.get("skipped", []),
        progress.get("failed", []),
    )

    runtime = build_runtime_summary(state)
    result = {
        "workflow_status": state.get("status"),
        "current_tasks": current,
        "total_tasks": total,
        "completed": len(progress.get("completed", [])),
        "failed": len(progress.get("failed", [])),
        "skipped": len(progress.get("skipped", [])),
        "progress_percent": pct,
        "progress_bar": generate_progress_bar(pct),
        **runtime,
    }

    if state.get("failure_reason"):
        result["failure_reason"] = state["failure_reason"]

    return result


def cmd_list(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """List all tasks with status."""
    state, _, tasks_content, _ = resolve_state_and_tasks(project_id, project_root)
    if not state or not tasks_content:
        return {"error": "没有活跃的工作流或任务"}

    tasks = parse_tasks_v2(tasks_content)
    return {
        "total": len(tasks),
        "tasks": [
            {
                "id": t.id,
                "name": t.name,
                "phase": t.phase,
                "status": t.status,
                "emoji": get_status_emoji(t.status),
                "quality_gate": t.quality_gate,
                "actions": t.actions,
            }
            for t in tasks
        ],
    }


def cmd_next(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """Find the next task to execute."""
    state, _, tasks_content, _ = resolve_state_and_tasks(project_id, project_root)
    if not state or not tasks_content:
        return {"error": "没有活跃的工作流或任务"}

    progress = state.get("progress", {})
    next_id = find_next_task(
        tasks_content,
        completed=progress.get("completed", []),
        skipped=progress.get("skipped", []),
        failed=progress.get("failed", []),
        blocked=progress.get("blocked", []),
    )

    if not next_id:
        return {"next_task": None, "message": "所有任务已完成或被阻塞"}

    tasks = parse_tasks_v2(tasks_content)
    task = next((t for t in tasks if t.id == next_id), None)
    if task:
        return {"next_task": task_to_dict(task)}
    return {"next_task": next_id}


def cmd_complete(
    task_id: str,
    project_id: Optional[str] = None,
    project_root: Optional[str] = None,
) -> Dict[str, Any]:
    """Mark a task as completed."""
    state, state_path, tasks_content, tasks_path = resolve_state_and_tasks(
        project_id, project_root
    )
    if not state or not tasks_content or not tasks_path or not state_path:
        return {"error": "没有活跃的工作流或任务"}

    # Update markdown
    updated = update_task_status_in_markdown(tasks_content, task_id, "completed")
    with open(tasks_path, "w", encoding="utf-8") as f:
        f.write(updated)

    # Update state
    progress = state.setdefault("progress", {})
    completed = progress.setdefault("completed", [])
    add_unique(completed, task_id)

    # Remove from failed if was retried
    if task_id in progress.get("failed", []):
        progress["failed"].remove(task_id)

    write_state(state_path, state)
    return {"completed": True, "task_id": task_id}


def cmd_fail(
    task_id: str,
    reason: str,
    project_id: Optional[str] = None,
    project_root: Optional[str] = None,
) -> Dict[str, Any]:
    """Mark a task as failed."""
    state, state_path, tasks_content, tasks_path = resolve_state_and_tasks(
        project_id, project_root
    )
    if not state or not tasks_content or not tasks_path or not state_path:
        return {"error": "没有活跃的工作流或任务"}

    updated = update_task_status_in_markdown(tasks_content, task_id, "failed")
    with open(tasks_path, "w", encoding="utf-8") as f:
        f.write(updated)

    state["status"] = "failed"
    state["failure_reason"] = reason
    state["current_tasks"] = [task_id]
    progress = state.setdefault("progress", {})
    add_unique(progress.setdefault("failed", []), task_id)
    write_state(state_path, state)

    return {"failed": True, "task_id": task_id, "reason": reason}


def cmd_deps(
    task_id: str,
    project_id: Optional[str] = None,
    project_root: Optional[str] = None,
) -> Dict[str, Any]:
    """Check task dependencies."""
    state, _, tasks_content, _ = resolve_state_and_tasks(project_id, project_root)
    if not state or not tasks_content:
        return {"error": "没有活跃的工作流或任务"}

    tasks = parse_tasks_v2(tasks_content)
    task = next((t for t in tasks if t.id == task_id), None)
    if not task:
        return {"error": f"任务 {task_id} 不存在"}

    progress = state.get("progress", {})
    result = check_task_deps(task.depends, progress.get("completed", []))
    result["task_id"] = task_id
    result["depends"] = task.depends
    result["blocked_by"] = task.blocked_by
    return result


def cmd_parallel(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """Find parallelizable task groups."""
    state, _, tasks_content, _ = resolve_state_and_tasks(project_id, project_root)
    if not state or not tasks_content:
        return {"error": "没有活跃的工作流或任务"}

    tasks = parse_tasks_v2(tasks_content)
    progress = state.get("progress", {})

    # Convert to dicts for dependency_checker
    task_dicts = [task_to_dict(t) for t in tasks]
    groups = find_parallel_groups(
        task_dicts,
        completed=progress.get("completed", []),
        blocked=progress.get("blocked", []),
        skipped=progress.get("skipped", []),
        failed=progress.get("failed", []),
    )

    return {"parallel_groups": groups, "group_count": len(groups)}


def cmd_progress(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """Show progress with visual bar."""
    state, _, tasks_content, _ = resolve_state_and_tasks(project_id, project_root)
    if not state or not tasks_content:
        return {"error": "没有活跃的工作流或任务"}

    progress = state.get("progress", {})
    total = count_tasks(tasks_content)
    pct = calculate_progress(
        total,
        progress.get("completed", []),
        progress.get("skipped", []),
        progress.get("failed", []),
    )

    return {
        "total": total,
        "completed": len(progress.get("completed", [])),
        "skipped": len(progress.get("skipped", [])),
        "failed": len(progress.get("failed", [])),
        "blocked": len(progress.get("blocked", [])),
        "pending": total - len(progress.get("completed", [])) - len(progress.get("skipped", [])) - len(progress.get("failed", [])),
        "percent": pct,
        "bar": generate_progress_bar(pct),
        "constraints": extract_constraints(tasks_content),
    }


def cmd_context_budget(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """Show context budget status."""
    state, _, _, _ = resolve_state_and_tasks(project_id, project_root)
    if not state:
        return {"error": "没有活跃的工作流"}

    metrics = state.get("contextMetrics", {})
    usage = metrics.get("usagePercent", 0)
    projected = metrics.get("projectedUsagePercent", 0)

    budget = evaluate_budget_thresholds(projected)
    budget["current_usage"] = usage
    budget["context_bar"] = generate_context_bar(usage)
    budget["max_consecutive_tasks"] = metrics.get("maxConsecutiveTasks", 5)
    budget["consecutive_count"] = state.get("consecutive_count", 0)

    return budget


def cmd_runtime_summary(
    project_id: Optional[str] = None, project_root: Optional[str] = None
) -> Dict[str, Any]:
    """聚合读侧 runtime 摘要，供 status/context/CLI 复用。"""
    state, _, _, _ = resolve_state_and_tasks(project_id, project_root)
    if not state:
        return {"error": "没有活跃的工作流"}

    return build_runtime_summary(state)


# =============================================================================
# CLI
# =============================================================================


def main() -> int:
    parser = argparse.ArgumentParser(description="任务管理 CLI")
    parser.add_argument("--project-id", help="项目 ID")
    parser.add_argument("--project-root", help="项目根目录（用于解析相对 plan_file）")
    sub = parser.add_subparsers(dest="command")

    sub.add_parser("status", help="当前状态")
    sub.add_parser("list", help="列出所有任务")
    sub.add_parser("next", help="下一个任务")

    p_complete = sub.add_parser("complete", help="标记完成")
    p_complete.add_argument("task_id", help="任务 ID")

    p_fail = sub.add_parser("fail", help="标记失败")
    p_fail.add_argument("task_id", help="任务 ID")
    p_fail.add_argument("reason", help="失败原因")

    p_deps = sub.add_parser("deps", help="检查依赖")
    p_deps.add_argument("task_id", help="任务 ID")

    sub.add_parser("parallel", help="查找可并行任务")
    sub.add_parser("progress", help="进度统计")
    sub.add_parser("context-budget", help="上下文预算")

    args = parser.parse_args()
    pid = args.project_id
    project_root = args.project_root

    commands = {
        "status": lambda: cmd_status(pid, project_root),
        "list": lambda: cmd_list(pid, project_root),
        "next": lambda: cmd_next(pid, project_root),
        "complete": lambda: cmd_complete(args.task_id, pid, project_root),
        "fail": lambda: cmd_fail(args.task_id, args.reason, pid, project_root),
        "deps": lambda: cmd_deps(args.task_id, pid, project_root),
        "parallel": lambda: cmd_parallel(pid, project_root),
        "progress": lambda: cmd_progress(pid, project_root),
        "context-budget": lambda: cmd_context_budget(pid, project_root),
    }

    handler = commands.get(args.command)
    if handler:
        result = handler()
        print(json.dumps(result, indent=2, ensure_ascii=False))
    else:
        parser.print_help()
        return 1

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: core/utils/workflow/task_manager.js
#!/usr/bin/env node
/** @file 任务管理器 - 工作流任务的状态查询、完成、失败、依赖检查、并行分组等命令 */

const fs = require('fs')
const path = require('path')
const { checkTaskDeps } = require('./dependency_checker')
const {
  detectProjectIdFromRoot,
  getWorkflowStatePath,
  resolveUnder,
  validateProjectId,
} = require('./path_utils')
const {
  calculateProgress,
  generateProgressBar,
  readState,
  writeState,
} = require('./state_manager')
const { addUnique, getStatusEmoji } = require('./status_utils')
const {
  extractConstraints,
} = require('./task_parser')
const { createTaskSource } = require('./task_source')
const { findOrphanedAnchors, firstDispatchableTaskId } = require('./workflow_types')

// S3 重基（FR-2）：task_manager 的 task 序列来源从 parseTasksV2(plan.md) 切到 TaskSource。
// 这些命令（complete/next/list/progress/deps）是 execute 主推进路径（workflow_cli advance）的真实落点。
// T9（C-7/C-1）：经 createTaskSource 工厂选 adapter——task-dir → TaskDirSource，仅 legacy plan.md → LegacyPlanMdSource。
function resolveProjectIdForTasks(state, projectId = null, projectRoot = null) {
  return projectId || (state && (state.project_id || state.projectId)) || detectProjectId(projectRoot) || null
}

// 取 TaskSource：legacy plan.md workflow 缺 task-dir 时回退 LegacyPlanMdSource（真实 advance 路径 C-7 兜底）。
// quiet=true：task_manager 内部多命令复用，迁移提示由首个命中处打印（进程内去重），避免重复刷屏。
function taskSourceFor(state, projectId = null, projectRoot = null) {
  return createTaskSource(state, { projectId, projectRoot, quiet: true })
}

// 下一个待执行 task：按 TaskSource 顺序排除 completed/skipped/failed/blocked。
// 实现收敛到 workflow_types.firstDispatchableTaskId（C-1 共享谓词），此处保留导出名兼容既有调用方。
function nextTaskIdFromSource(tasks, progress = {}) {
  return firstDispatchableTaskId(tasks, progress)
}

/**
 * 检测当前项目的项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录路径
 * @returns {string|null} 项目 ID
 */
function detectProjectId(projectRoot = null) {
  return detectProjectIdFromRoot(projectRoot)
}

/**
 * 检测并解析项目根目录的绝对路径
 * @param {string|null} [projectRoot=null] - 项目根目录路径，为空时自动检测
 * @returns {string} 解析后的项目根目录绝对路径
 */
function detectProjectRoot(projectRoot = null) {
  if (projectRoot) return path.resolve(projectRoot)
  const configPath = path.join(process.cwd(), '.claude', 'config', 'project-config.json')
  if (fs.existsSync(configPath)) return path.dirname(path.dirname(path.dirname(configPath)))
  return process.cwd()
}

/**
 * 解析 Plan 产物的绝对路径
 * @param {string} projectRoot - 项目根目录
 * @param {string} artifactRef - 产物引用路径（相对或绝对）
 * @returns {string|null} 解析后的绝对路径，无效时返回 null
 */
function resolvePlanArtifactPath(projectRoot, artifactRef) {
  if (!artifactRef) return null
  if (path.isAbsolute(artifactRef)) return artifactRef
  const resolved = resolveUnder(projectRoot, artifactRef)
  if (resolved) return resolved
  const fallback = path.resolve(projectRoot, artifactRef)
  const projectRootResolved = path.resolve(projectRoot)
  return fallback === projectRootResolved || fallback.startsWith(`${projectRootResolved}${path.sep}`) ? fallback : null
}

/**
 * 解析工作流状态和任务内容。
 * @returns {Array} [state, statePath, tasksContent, tasksPath, code]
 *   - 成功：前 4 项有效，code === null
 *   - 失败：未解析的项为 null，code 为 snake_case 根因
 *     （project_id_missing / project_id_invalid / state_file_missing /
 *      plan_file_unset / plan_path_unresolved / plan_file_missing）
 */
function resolveStateAndTasks(projectId = null, projectRoot = null) {
  const pid = projectId || detectProjectId(projectRoot)
  if (!pid) return [null, null, null, null, 'project_id_missing']
  if (!validateProjectId(pid)) return [null, null, null, null, 'project_id_invalid']
  const statePath = getWorkflowStatePath(pid)
  if (!statePath || !fs.existsSync(statePath)) return [null, null, null, null, 'state_file_missing']
  const state = readState(statePath, pid)
  const planRef = state.plan_file || ''
  if (!planRef) return [state, statePath, null, null, 'plan_file_unset']
  const resolvedProjectRoot = detectProjectRoot(projectRoot || state.project_root)
  const artifactPath = resolvePlanArtifactPath(resolvedProjectRoot, planRef)
  if (!artifactPath) return [state, statePath, null, null, 'plan_path_unresolved']
  if (!fs.existsSync(artifactPath)) return [state, statePath, null, null, 'plan_file_missing']
  return [state, statePath, fs.readFileSync(artifactPath, 'utf8'), artifactPath, null]
}

/**
 * 首次推进时把 state.status 从 planned 升为 running。
 * 返回 'planned->running' 以便调用方把信号带回调用方；其他状态不动并返回 null。
 */
function liftPlannedToRunning(state) {
  if (state.status !== 'planned') return null
  state.status = 'running'
  return 'planned->running'
}

/**
 * 构建工作流运行时摘要，汇总 delta 追踪、规划门控、上下文注入等信息
 * @param {Object} state - 工作流状态对象
 * @returns {Object} 运行时摘要对象
 */
function buildRuntimeSummary(state) {
  const reviewStatus = state.review_status || {}
  return {
    delta_tracking: state.delta_tracking || {},
    planning_gates: {
      discussion: state.discussion || {},
      requirement_baseline: state.requirement_baseline || {},
      ux_design: state.ux_design || {},
      user_spec_review: reviewStatus.user_spec_review || {},
      plan_review: reviewStatus.plan_review || {},
    },
    context_injection: state.context_injection || {},
    unblocked: state.unblocked || [],
  }
}

/**
 * 查询工作流状态概览
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 状态概览，包含进度、任务数、运行时摘要等
 */
function cmdStatus(projectId = null, projectRoot = null) {
  const [state, , , tasksPath, code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流', code }
  const progress = state.progress || {}
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  const source = taskSourceFor(state, pid, projectRoot)
  const sourceTasks = source ? source.listTasks() : []
  const total = sourceTasks.length
  const percent = calculateProgress(total, progress.completed || [], progress.skipped || [], progress.failed || [])
  // C-1 锚点可解析性只读暴露：current_tasks 含不在 task 源的 id → 派生 current_tasks_orphaned 标记，
  // 让 status 直接呈现矛盾（否则 current_tasks 原样回显 + total_tasks 各自健康，失配被双重掩盖）。
  // 全量 current_tasks 检查（对齐 plan-review 广度）；复用 listTasks 结果免二次读盘。
  const anchorOrphaned = Boolean(source) && findOrphanedAnchors(state.current_tasks, sourceTasks).length > 0
  return {
    workflow_status: state.status,
    plan_file: tasksPath || state.plan_file || '',
    current_tasks: state.current_tasks || [],
    ...(anchorOrphaned ? { current_tasks_orphaned: true } : {}),
    total_tasks: total,
    completed: (progress.completed || []).length,
    failed: (progress.failed || []).length,
    skipped: (progress.skipped || []).length,
    progress_percent: percent,
    progress_bar: generateProgressBar(percent),
    ...buildRuntimeSummary(state),
    ...(state.failure_reason ? { failure_reason: state.failure_reason } : {}),
  }
}

/**
 * 列出所有任务及其状态
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 任务列表，包含 total 和 tasks 数组
 */
function cmdList(projectId = null, projectRoot = null) {
  const [state, , , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  const source = taskSourceFor(state, pid, projectRoot)
  const tasks = source ? source.listTasks() : []
  return {
    total: tasks.length,
    tasks: tasks.map((task) => ({
      id: task.id,
      name: task.name || '',
      phase: task.phase,
      status: task.status,
      emoji: getStatusEmoji(task.status),
      target_layer: task.target_layer,
      package: task.package,
    })),
  }
}

/**
 * 查找下一个待执行的任务
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 下一个任务信息或完成提示
 */
function cmdNext(projectId = null, projectRoot = null) {
  const [state, , , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  const source = taskSourceFor(state, pid, projectRoot)
  const tasks = source ? source.listTasks() : []
  // task 源为空 + plan 未解析 → 透传 resolveStateAndTasks 的诊断 code（如 plan_file_unset / plan_file_missing），
  // 保持「无任务」的可诊断契约（task-dir 流程任务齐时 code 为 null，不影响正常推进）。
  if (!tasks.length && code) return { error: '没有活跃的工作流或任务', code }
  const progress = state.progress || {}
  const nextId = nextTaskIdFromSource(tasks, progress)
  if (!nextId) return { next_task: null, message: '所有任务已完成或被阻塞' }
  const task = source ? source.getTask(nextId) : null
  return { next_task: task || nextId }
}

/**
 * 将指定任务标记为已完成
 * @param {string} taskId - 任务 ID
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 操作结果
 */
function cmdComplete(taskId, projectId = null, projectRoot = null) {
  const [state, statePath, , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state || !statePath) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  if (!pid) return { error: '无法解析 project id，无法标记完成', code }
  const source = taskSourceFor(state, pid, projectRoot)
  const task = source ? source.getTask(taskId) : null
  if (!task) return { error: `任务 ${taskId} 不存在于 task 源中，无法标记完成` }
  source.updateTaskStatus(taskId, 'completed')
  const progress = state.progress || (state.progress = {})
  const completed = progress.completed || (progress.completed = [])
  addUnique(completed, taskId)
  // complete 终结该 task：failed/blocked 一并清除（与 markTaskSkipped 对称）——
  // 终结 id 残留 blocked 会污染 selectAnchorId 回退域，让 repair-anchor 面对本可避免的脏数据。
  if ((progress.failed || []).includes(taskId)) progress.failed = progress.failed.filter((item) => item !== taskId)
  if ((progress.blocked || []).includes(taskId)) progress.blocked = progress.blocked.filter((item) => item !== taskId)
  const statusTransition = liftPlannedToRunning(state)
  writeState(statePath, state)
  const result = { completed: true, task_id: taskId }
  if (statusTransition) result.status_transition = statusTransition
  return result
}

/**
 * 将指定任务标记为失败
 * @param {string} taskId - 任务 ID
 * @param {string} reason - 失败原因
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 操作结果
 */
function cmdFail(taskId, reason, projectId = null, projectRoot = null) {
  const [state, statePath, , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state || !statePath) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  if (!pid) return { error: '无法解析 project id，无法标记失败', code }
  const source = taskSourceFor(state, pid, projectRoot)
  const task = source ? source.getTask(taskId) : null
  if (!task) return { error: `任务 ${taskId} 不存在于 task 源中，无法标记失败` }
  source.updateTaskStatus(taskId, 'failed')
  state.status = 'halted'
  state.halt_reason = 'failure'
  state.failure_reason = reason
  state.current_tasks = [taskId]
  const progress = state.progress || (state.progress = {})
  const failed = progress.failed || (progress.failed = [])
  addUnique(failed, taskId)
  // 与 cmdComplete 对称：失败时把 taskId 移出 completed，避免 completed→fail 同 task 双计入。
  if ((progress.completed || []).includes(taskId)) progress.completed = progress.completed.filter((item) => item !== taskId)
  writeState(statePath, state)
  return { failed: true, task_id: taskId, reason }
}

/**
 * 查询指定任务的依赖状态
 * @param {string} taskId - 任务 ID
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 依赖检查结果
 */
function cmdDeps(taskId, projectId = null, projectRoot = null) {
  const [state, , , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  const source = taskSourceFor(state, pid, projectRoot)
  const task = source ? source.getTask(taskId) : null
  if (!task) return { error: `任务 ${taskId} 不存在` }
  const progress = state.progress || {}
  // task-dir 记录仅含 depends（无 blocked_by）；blocked_by 为 plan.md 时代字段，task-dir 流程不再承载。
  return {
    ...checkTaskDeps(task.depends, progress.completed || []),
    task_id: taskId,
    depends: task.depends,
    blocked_by: task.blocked_by || [],
  }
}

/**
 * 查询工作流整体进度
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 进度详情，包含完成数、失败数、百分比、进度条等
 */
function cmdProgress(projectId = null, projectRoot = null) {
  const [state, , tasksContent, , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流或任务', code }
  const pid = resolveProjectIdForTasks(state, projectId, projectRoot)
  const source = taskSourceFor(state, pid, projectRoot)
  const progress = state.progress || {}
  const total = source ? source.listTasks().length : 0
  const percent = calculateProgress(total, progress.completed || [], progress.skipped || [], progress.failed || [])
  return {
    total,
    completed: (progress.completed || []).length,
    skipped: (progress.skipped || []).length,
    failed: (progress.failed || []).length,
    blocked: (progress.blocked || []).length,
    pending: total - (progress.completed || []).length - (progress.skipped || []).length - (progress.failed || []).length,
    percent,
    bar: generateProgressBar(percent),
    // 约束仍从 plan.md 叙述正文提取（如可读）；task-dir 不承载约束文本。
    constraints: tasksContent ? extractConstraints(tasksContent) : [],
  }
}

/**
 * 获取工作流运行时摘要
 * @param {string|null} [projectId=null] - 项目 ID
 * @param {string|null} [projectRoot=null] - 项目根目录
 * @returns {Object} 运行时摘要
 */
function cmdRuntimeSummary(projectId = null, projectRoot = null) {
  const [state, , , , code] = resolveStateAndTasks(projectId, projectRoot)
  if (!state) return { error: '没有活跃的工作流', code }
  return buildRuntimeSummary(state)
}

function parseArgs(argv) {
  const args = [...argv]
  const options = { projectId: null, projectRoot: null }
  while (args.length && args[0].startsWith('--')) {
    const flag = args.shift()
    if (flag === '--project-id') options.projectId = args.shift()
    else if (flag === '--project-root') options.projectRoot = args.shift()
    else throw new Error(`Unknown flag: ${flag}`)
  }
  return { options, command: args.shift(), args }
}

function main() {
  try {
    const { options, command, args } = parseArgs(process.argv.slice(2))
    const handlers = {
      status: () => cmdStatus(options.projectId, options.projectRoot),
      list: () => cmdList(options.projectId, options.projectRoot),
      next: () => cmdNext(options.projectId, options.projectRoot),
      complete: () => cmdComplete(args[0], options.projectId, options.projectRoot),
      fail: () => cmdFail(args[0], args[1], options.projectId, options.projectRoot),
      deps: () => cmdDeps(args[0], options.projectId, options.projectRoot),
      progress: () => cmdProgress(options.projectId, options.projectRoot),
    }
    const handler = handlers[command]
    if (!handler) {
      process.stderr.write('Usage: node task_manager.js [--project-id ID] [--project-root DIR] <status|list|next|complete|fail|deps|progress> ...\n')
      process.exitCode = 1
      return
    }
    process.stdout.write(`${JSON.stringify(handler(), null, 2)}\n`)
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`)
    process.exitCode = 1
  }
}

module.exports = {
  detectProjectId,
  detectProjectRoot,
  resolvePlanArtifactPath,
  resolveStateAndTasks,
  liftPlannedToRunning,
  buildRuntimeSummary,
  cmdStatus,
  cmdList,
  cmdNext,
  cmdComplete,
  cmdFail,
  cmdDeps,
  cmdProgress,
  cmdRuntimeSummary,
}

if (require.main === module) main()
